### PR TITLE
CFN Graph edges

### PR DESCRIPTION
**Pull request by Saarett**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

1. Added CFN Edges for the functions Connection, Fn::If, Fn::FindInMap, Fn::Sub, Fn::GetAtt, and the resource attribute DependsOn
2. Ref connections are currently not created due to conflicts with default ref evaluation in create_definitions function
3. Created a CFN keywords file currently with IntrinsicFunctions, ConditionFunctions, ResourceAttributes and copied the CloudformationTemplateSections ENUM into this file (refactor)
4. Added tests

### CODE REVIEW

**Review by nimrodkor — COMMENTED** [review]
@Saarett Very nice! a few comments

File: checkov/cloudformation/graph_builder/local_graph.py [hunk 1]
```diff
@@ -62,5 +87,180 @@ def _create_section_vertices(self, file_path: str, file_conf: dict, section: Clo
...
+                sub_parameter_values = {}
+                source_id, value, attributes = self._extract_source_value_attrs(sub_obj)
+
+                if isinstance(value, list):
+                    if not value:
+                        continue
+                    if len(value) == 2:
+                        sub_parameter_values = value[1]
+                    sub_parameters = self._find_fn_sub_parameter(value[0])
+                elif isinstance(value, (six.text_type, six.string_types)):
```
Comment (gruebel) [inline checkov/cloudformation/graph_builder/local_graph.py:197]:
```suggestion
                elif isinstance(value, str):
```

File: checkov/cloudformation/graph_builder/local_graph.py [hunk 2]
```diff
@@ -62,5 +87,180 @@ def _create_section_vertices(self, file_path: str, file_conf: dict, section: Clo
...
+
+                for sub_parameter in sub_parameters:
+                    if sub_parameter not in sub_parameter_values:
+                        if '.' in sub_parameter:
+                            sub_parameter = sub_parameter.split('.')[0]
+                        origin_vertex_index, dest_vertex_index, label = self._extract_origin_dest_label(
+                            file_path, source_id, sub_parameter, attributes)
+                        if origin_vertex_index is None or dest_vertex_index is None:
+                            continue
+                        self._create_edge(origin_vertex_index, dest_vertex_index, label)
```
Comment (gruebel) [inline checkov/cloudformation/graph_builder/local_graph.py:208]:
```suggestion
                        if origin_vertex_index is not None and dest_vertex_index is not None:
                            self._create_edge(origin_vertex_index, dest_vertex_index, label)
```
same as @nimrodkor suggested before, looks cleaner 🙂

File: checkov/cloudformation/graph_builder/local_graph.py [hunk 3]
```diff
@@ -62,5 +87,180 @@ def _create_section_vertices(self, file_path: str, file_conf: dict, section: Clo
...
+            origin_vertex_index = dpath.get(self._vertices_indexes, [file_path, source_id])
+        if dpath.search(self._vertices_indexes, [file_path, target_id]):
+            dest_vertex_index = dpath.get(self._vertices_indexes, [file_path, target_id])
+        attributes_joined = '.'.join(map(str, attributes))  # mapping all attributes to str because one of the attrs might be an int
+        return origin_vertex_index, dest_vertex_index, attributes_joined
+
+    @staticmethod
+    def _find_fn_sub_parameter(string):
+        """Search string for tokenized fields"""
+        regex = re.compile(r'\${([a-zA-Z0-9.]*)}')
```
Comment (gruebel) [inline checkov/cloudformation/graph_builder/local_graph.py:222]: this wouldn't find pseudo parameters https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/pseudo-parameter-reference.html, but maybe that's also intended.

File: checkov/cloudformation/graph_builder/local_graph.py [hunk 4]
```diff
@@ -62,5 +87,180 @@ def _create_section_vertices(self, file_path: str, file_conf: dict, section: Clo
...
+        return regex.findall(string)
+
+    def _create_edges(self) -> None:
+        self._add_resource_attr_connections(ResourceAttributes.DEPENDS_ON.value)
+        self._add_resource_attr_connections(IntrinsicFunctions.CONDITION.value)
+        self._add_fn_connections(IntrinsicFunctions.GET_ATT.value)
+        self._add_fn_connections(ConditionFunctions.IF.value)
+        #self._add_fn_connections(IntrinsicFunctions.REF.value)
```
Comment (gruebel) [inline checkov/cloudformation/graph_builder/local_graph.py:230]: any reason, why it is commented?

File: checkov/cloudformation/parser/__init__.py [hunk 5]
```diff
@@ -43,4 +44,11 @@ def parse(filename: str) -> Union[Tuple[dict_node, List[Tuple[int, str]]], Tuple
     except YAMLError as err:
         pass
 
+    if template:
+        resources = template.get(TemplateSections.RESOURCES.value, None)
+        if resources:
+            if '__startline__' in resources:
+                resources.pop('__startline__')
+            if '__endline__' in resources:
+                resources.pop('__endline__')
```
Comment (gruebel) [inline checkov/cloudformation/parser/__init__.py:53]:
```suggestion
            if '__startline__' in resources:
                del resources['__startline__']
            if '__endline__' in resources:
                del resources['__endline__']
```
is performance wise better, `pop` is usually used to actually do something with the removed key-value or you could remove the `if` condition and directly use `pop`, like `resources.pop('__startline__', None)`

File: checkov/cloudformation/graph_builder/local_graph.py (hunk 4 above)
Comment (Saarett) [reply checkov/cloudformation/graph_builder/local_graph.py:230]: Conflicts with a default ref evaluation done prior in another function in CFN runner and @nimrodkor fixed in another PR

File: checkov/cloudformation/graph_builder/local_graph.py (hunk 3 above)
Comment (Saarett) [reply checkov/cloudformation/graph_builder/local_graph.py:222]: That's the intention, thanks :)

File: checkov/cloudformation/runner.py [hunk 6]
```diff
@@ -6,7 +6,7 @@
 from checkov.cloudformation.cfn_utils import create_definitions, build_definitions_context
 from checkov.cloudformation.checks.resource.registry import cfn_registry
 from checkov.cloudformation.context_parser import ContextParser
-from checkov.cloudformation.graph_builder.graph_components.block_types import CloudformationTemplateSections
+from checkov.cloudformation.parser import TemplateSections
```
[the remark this answers is not in the document]
Comment (Saarett) [reply checkov/cloudformation/runner.py:9]: @nimrodkor I don't mind putting the cfn_keywords file in another package so runner could use it. Where do you think is the appropriate place?

[the remark this answers is not in the document]
Comment (nimrodkor) [reply checkov/cloudformation/runner.py:9]: Maybe this is a bit better:
```suggestion
from checkov.cloudformation.parser.cfn_keywords import TemplateSections
```

File: checkov/cloudformation/graph_builder/local_graph.py (hunk 4 above)
Comment (nimrodkor) [reply checkov/cloudformation/graph_builder/local_graph.py:230]: See #1501

**Review by nimrodkor — APPROVED** [review]
@Saarett Very well done!

**Pull request merged**